cmd/bosun/expr/funcs.go

```diff
@@ -440,6 +447,86 @@ func Filter(e *State, T miniprofiler.Timer, series *Results, number *Results) (*
 	return series, nil
 }
 
+func quickSelect(data sort.Interface, k int, max bool) {
+
+	var pivotIndex int
+	low := 0
+	high := data.Len() - 1
+
+	// if we asked to last k and we
+	// don't have that much everyone
+	// is in the set
+	if k > data.Len() {
+		return
+	}
+
+	for {
+
+		if low >= high {
+			return
+		}
+
+		pivotIndex = rand.Intn(high+1-low) + low
+		pivotIndex = partition(data, low, high, pivotIndex, max)
+
+		if k < pivotIndex {
+			high = pivotIndex - 1
+		} else if k > pivotIndex {
+			low = pivotIndex + 1
+		} else {
+			return
+		}
+	}
+}
+
+func partition(data sort.Interface, low, high, pivotIndex int, max bool) int {
+	partitionIndex := low
+	data.Swap(pivotIndex, high)
+	for i := low; i < high; i++ {
+		if (!data.Less(i, high) && max) || (data.Less(i, high) && !max) {
+			data.Swap(i, partitionIndex)
+			partitionIndex++
+		}
+
+	}
+	data.Swap(partitionIndex, high)
+	return partitionIndex
+}
+
+func Tail(e *State, T miniprofiler.Timer, series *Results, v float64) (*Results, error) {
+	for _, sr := range series.Results {
+		// if there are fewer points than the requested tail
+		// short circut and just return current series
+		if len(sr.Value.Value().(Series)) > int(v) {
```

Comment:
invert condition and continue so most of code is not indented - i.e. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow